find_library linkable check is too strict for RTEMS

**Describe the bug**
In an existing repo where build was working, upgrading to meson 1.10 breaks `meson setup`

```
-----------
Command line: `/home/michael/sdk/rtems-noel-1.0.4/2022.06/bin/riscv-rtems5-gcc -I/home/michael/sdk/rtems-noel-1.0.4/2022.06/bin/../lib/gcc/riscv-rtems5/9.3.0/include -I/home/michael/sdk/rtems-noel-1.0.4/2022.06/bin/../lib/gcc/riscv-rtems5/9.3.0/include-fixed -I/home/michael/sdk/rtems-noel-1.0.4/2022.06/bin/../lib/gcc/riscv-rtems5/9.3.0/../../../../riscv-rtems5/include -I/home/michael/sdk/rtems-noel-1.0.4/2022.06/riscv-rtems5/noel32imafd/lib/include/ -I/home/michael/sdk/rtems-noel-1.0.4/2022.06/kernel/riscv-rtems5/noel32imafd/lib/include/ /home/michael/proj/pfgen/endurance-flight-software/cmake/build/meson-private/tmpxn87jx67/testfile.c -o /home/michael/proj/pfgen/endurance-flight-software/cmake/build/meson-private/tmpxn87jx67/output.exe -B/home/michael/sdk/rtems-noel-1.0.4/2022.06/riscv-rtems5/noel32imafd/lib -B/home/michael/sdk/rtems-noel-1.0.4/2022.06/kernel/riscv-rtems5/noel32imafd/lib -Wall -Wextra -fno-common -ffunction-sections -fdata-sections -D_RTEMS_5_ -U_FILE_OFFSET_BITS -qrtems -Wfatal-errors -Werror=int-conversion -Werror=format -Werror=format-nonliteral -Werror=format-signedness -Werror=nonnull -Werror=implicit -Werror=implicit-fallthrough -Werror=missing-braces -Werror=incompatible-pointer-types -Werror=overflow -D_FILE_OFFSET_BITS=64 -O0 -Wl,--start-group /home/michael/proj/pfgen/endurance-flight-software/cmake/karvel/modules/karvel_endurance/lib/libkarvel_endurance.a -Wl,--allow-shlib-undefined /home/michael/sdk/rtems-noel-1.0.4/2022.06/bin/../lib/gcc/riscv-rtems5/9.3.0/crti.o /home/michael/sdk/rtems-noel-1.0.4/2022.06/bin/../lib/gcc/riscv-rtems5/9.3.0/crtbegin.o -nostdlib -nodefaultlibs -lrtemscpu -lrtemsbsp -lgcc -lc -Wl,--end-group -Wl,--gc-sections /home/michael/sdk/rtems-noel-1.0.4/2022.06/bin/../lib/gcc/riscv-rtems5/9.3.0/crtend.o /home/michael/sdk/rtems-noel-1.0.4/2022.06/bin/../lib/gcc/riscv-rtems5/9.3.0/crtn.o -T/home/michael/sdk/rtems-noel-1.0.4/2022.06/kernel/riscv-rtems5/noel32imafd/lib/linkcmds` -> 1
stderr:
/home/michael/sdk/rtems-noel-1.0.4/2022.06/bin/../lib/gcc/riscv-rtems5/9.3.0/../../../../riscv-rtems5/bin/ld: /home/michael/sdk/rtems-noel-1.0.4/2022.06/kernel/riscv-rtems5/noel32imafd/lib/start.o: in function `.L0 ':
/home/jernej/repositories/rsb/build-toolchain/src/bsp/riscv-rtems5/c/noel32imafd/lib/libbsp/riscv/riscv/../../../../../../../../rtems-noel-1.0.4/c/src/lib/libbsp/riscv/riscv/../../../../../../bsps/riscv/shared/start/start.S:77: undefined reference to `_ISR_Stack_area_end'
/home/michael/sdk/rtems-noel-1.0.4/2022.06/bin/../lib/gcc/riscv-rtems5/9.3.0/../../../../riscv-rtems5/bin/ld: /home/michael/sdk/rtems-noel-1.0.4/2022.06/kernel/riscv-rtems5/noel32imafd/lib/librtemscpu.a(userextiterate.o): in function `_User_extensions_Iterate':
/home/jernej/repositories/rsb/build-toolchain/src/bsp/riscv-rtems5/c/noel32imafd/cpukit/../../../../../rtems-noel-1.0.4/c/src/../../cpukit/score/src/userextiterate.c:167: undefined reference to `_User_extensions_Initial_count'

[...]

/home/michael/sdk/rtems-noel-1.0.4/2022.06/bin/../lib/gcc/riscv-rtems5/9.3.0/../../../../riscv-rtems5/bin/ld: /home/michael/sdk/rtems-noel-1.0.4/2022.06/kernel/riscv-rtems5/noel32imafd/lib/librtemscpu.a(ftruncate.o): in function `rtems_libio_iop':
/home/jernej/repositories/rsb/build-toolchain/src/rtems-noel-1.0.4/cpukit/include/rtems/libio_.h:130: undefined reference to `rtems_libio_iops'
/home/michael/sdk/rtems-noel-1.0.4/2022.06/bin/../lib/gcc/riscv-rtems5/9.3.0/../../../../riscv-rtems5/bin/ld: /home/jernej/repositories/rsb/build-toolchain/src/rtems-noel-1.0.4/cpukit/include/rtems/libio_.h:130: undefined reference to `rtems_libio_iops'
collect2: error: ld returned 1 exit status
-----------

karvel/meson.build:190:11: ERROR: C static library 'karvel_endurance' not found
```

**To Reproduce**

1. `uv tool install meson@1.9.2`
2. `meson setup build/ --cross cross_riscv-rtems5-noel32imafd.ini`
3. Setup works 
4. `uv tool install meson@1.10.1`
5. `rm -Rf build/`
6. `meson setup build/ --cross cross_riscv-rtems5-noel32imafd.ini`
7. Setup fails

I have attached the meson-logs.txt for each case and the cross file I am using

The relevant part of the meson.build file are:

```meson
# Binary libraries
karvel_bin_lib = [
  'modules/karvel_endurance',
]

cc = meson.get_compiler('c')
root = meson.current_source_dir()

# We deal with binary modules first, since source modules are likely to depend on them
foreach l : karvel_bin_lib

  # Fetch variables from subdir
  subdir(l)

  lib = cc.find_library(lib_name, dirs: [f'@root@/@l@/lib'], required: true, static: true)
```

[meson-log.1.9.2.txt](https://github.com/user-attachments/files/25799802/meson-log.1.9.2.txt)
[meson-log.1.10.1.txt](https://github.com/user-attachments/files/25799801/meson-log.1.10.1.txt)
[cross_riscv-rtems5-noel32imafd.ini.txt](https://github.com/user-attachments/files/25799806/cross_riscv-rtems5-noel32imafd.ini.txt)

**Expected behavior**
Setup should work as before

**system parameters**
* Cross-build
* OS: Ubuntu 24.04 LTS
* Python 3.13.3
* Meson 1.10.1
* Ninja 1.13.0

In general it is advisable to include an excerpt of the error such that reading the initial bug report does not require downloading a text file attachment with the real bug report. Let me post your build error for you. :)

```
-----------
Command line: `/home/michael/sdk/rtems-noel-1.0.4/2022.06/bin/riscv-rtems5-gcc -I/home/michael/sdk/rtems-noel-1.0.4/2022.06/bin/../lib/gcc/riscv-rtems5/9.3.0/include -I/home/michael/sdk/rtems-noel-1.0.4/2022.06/bin/../lib/gcc/riscv-rtems5/9.3.0/include-fixed -I/home/michael/sdk/rtems-noel-1.0.4/2022.06/bin/../lib/gcc/riscv-rtems5/9.3.0/../../../../riscv-rtems5/include -I/home/michael/sdk/rtems-noel-1.0.4/2022.06/riscv-rtems5/noel32imafd/lib/include/ -I/home/michael/sdk/rtems-noel-1.0.4/2022.06/kernel/riscv-rtems5/noel32imafd/lib/include/ /home/michael/proj/pfgen/endurance-flight-software/cmake/build/meson-private/tmpxn87jx67/testfile.c -o /home/michael/proj/pfgen/endurance-flight-software/cmake/build/meson-private/tmpxn87jx67/output.exe -B/home/michael/sdk/rtems-noel-1.0.4/2022.06/riscv-rtems5/noel32imafd/lib -B/home/michael/sdk/rtems-noel-1.0.4/2022.06/kernel/riscv-rtems5/noel32imafd/lib -Wall -Wextra -fno-common -ffunction-sections -fdata-sections -D_RTEMS_5_ -U_FILE_OFFSET_BITS -qrtems -Wfatal-errors -Werror=int-conversion -Werror=format -Werror=format-nonliteral -Werror=format-signedness -Werror=nonnull -Werror=implicit -Werror=implicit-fallthrough -Werror=missing-braces -Werror=incompatible-pointer-types -Werror=overflow -D_FILE_OFFSET_BITS=64 -O0 -Wl,--start-group /home/michael/proj/pfgen/endurance-flight-software/cmake/karvel/modules/karvel_endurance/lib/libkarvel_endurance.a -Wl,--allow-shlib-undefined /home/michael/sdk/rtems-noel-1.0.4/2022.06/bin/../lib/gcc/riscv-rtems5/9.3.0/crti.o /home/michael/sdk/rtems-noel-1.0.4/2022.06/bin/../lib/gcc/riscv-rtems5/9.3.0/crtbegin.o -nostdlib -nodefaultlibs -lrtemscpu -lrtemsbsp -lgcc -lc -Wl,--end-group -Wl,--gc-sections /home/michael/sdk/rtems-noel-1.0.4/2022.06/bin/../lib/gcc/riscv-rtems5/9.3.0/crtend.o /home/michael/sdk/rtems-noel-1.0.4/2022.06/bin/../lib/gcc/riscv-rtems5/9.3.0/crtn.o -T/home/michael/sdk/rtems-noel-1.0.4/2022.06/kernel/riscv-rtems5/noel32imafd/lib/linkcmds` -> 1
stderr:
/home/michael/sdk/rtems-noel-1.0.4/2022.06/bin/../lib/gcc/riscv-rtems5/9.3.0/../../../../riscv-rtems5/bin/ld: /home/michael/sdk/rtems-noel-1.0.4/2022.06/kernel/riscv-rtems5/noel32imafd/lib/start.o: in function `.L0 ':
/home/jernej/repositories/rsb/build-toolchain/src/bsp/riscv-rtems5/c/noel32imafd/lib/libbsp/riscv/riscv/../../../../../../../../rtems-noel-1.0.4/c/src/lib/libbsp/riscv/riscv/../../../../../../bsps/riscv/shared/start/start.S:77: undefined reference to `_ISR_Stack_area_end'
/home/michael/sdk/rtems-noel-1.0.4/2022.06/bin/../lib/gcc/riscv-rtems5/9.3.0/../../../../riscv-rtems5/bin/ld: /home/michael/sdk/rtems-noel-1.0.4/2022.06/kernel/riscv-rtems5/noel32imafd/lib/librtemscpu.a(userextiterate.o): in function `_User_extensions_Iterate':
/home/jernej/repositories/rsb/build-toolchain/src/bsp/riscv-rtems5/c/noel32imafd/cpukit/../../../../../rtems-noel-1.0.4/c/src/../../cpukit/score/src/userextiterate.c:167: undefined reference to `_User_extensions_Initial_count'

[...]

/home/michael/sdk/rtems-noel-1.0.4/2022.06/bin/../lib/gcc/riscv-rtems5/9.3.0/../../../../riscv-rtems5/bin/ld: /home/michael/sdk/rtems-noel-1.0.4/2022.06/kernel/riscv-rtems5/noel32imafd/lib/librtemscpu.a(ftruncate.o): in function `rtems_libio_iop':
/home/jernej/repositories/rsb/build-toolchain/src/rtems-noel-1.0.4/cpukit/include/rtems/libio_.h:130: undefined reference to `rtems_libio_iops'
/home/michael/sdk/rtems-noel-1.0.4/2022.06/bin/../lib/gcc/riscv-rtems5/9.3.0/../../../../riscv-rtems5/bin/ld: /home/jernej/repositories/rsb/build-toolchain/src/rtems-noel-1.0.4/cpukit/include/rtems/libio_.h:130: undefined reference to `rtems_libio_iops'
collect2: error: ld returned 1 exit status
-----------

karvel/meson.build:190:11: ERROR: C static library 'karvel_endurance' not found
```

--- a/mesonbuild/compilers/mixins/clike.py
+++ b/mesonbuild/compilers/mixins/clike.py
@@ -1149,11 +1149,45 @@ class CLikeCompiler(Compiler):
                 for trial in trials:
                     if not os.path.isfile(trial):
                         continue
+
+                    # Check if an argument to find_library is usable:
+                    # * user requested a static library specifically
+                    #   => assume they know what they're doing, no link test
+                    #      as some runtime libraries may be provided in such
+                    #      a way that they must be used together with another
+                    #      library, or require special options.
+                    #
+                    # * user requested a library without `static: true`
+                    #   => build a shared library and link against it. Shared
+                    #      libraries (on ELF platforms) can be underlinked
+                    #      so it's better than an executable, as we'll only fail
+                    #      if it's unusable for another reason.
+
+                    # Don't bother fishing for arguments if we've already
+                    # been told not to perform a test.
+                    if not skip_link_check:
+                        shared_link_args = self.get_std_shared_lib_link_args()
+                        extra_args = [trial] + lcargs + shared_link_args
+
+                        # Try to build the link test as PIC. We don't know if
+                        # the target library is itself PIC: if it's non-PIC,
+                        # PIC is fine, and if it is PIC, then PIC is needed,
+                        # so always use it.
+                        extra_args += self.get_pic_args()
+
+                        # If we were asked specifically for a static library, don't perform
+                        # a link test, because it might be a special toolchain library like
+                        # on RTEMS.
+                        skip_link_check |= (libtype == LibType.STATIC)
+
+                        # Don't bother with a link check if we don't know how to build
+                        # a shared library.
+                        skip_link_check |= (not shared_link_args)
+
                     # When skip_link_check is True (e.g. for pkg-config
                     # libraries which are trusted to be linkable), skip the
                     # potentially expensive link check and just verify that the
                     # file exists.
-                    extra_args = [trial] + lcargs
                     if skip_link_check or self.links(code, extra_args=extra_args, disable_cache=True)[0]:
                         trial_result = trial
                         break

--- a/mesonbuild/compilers/mixins/clike.py
+++ b/mesonbuild/compilers/mixins/clike.py
@@ -1149,11 +1149,44 @@ class CLikeCompiler(Compiler):
                 for trial in trials:
                     if not os.path.isfile(trial):
                         continue
+
+                    # Check if an argument to find_library is usable:
+                    # * user requested a static library specifically
+                    #   => assume they know what they're doing, no link test
+                    #      as some runtime libraries may be provided in such
+                    #      a way that they must be used together with another
+                    #      library, or require special options.
+                    #
+                    # * user requested a library without `static: true`
+                    #   => build a shared library and link against it. Shared
+                    #      libraries (on ELF platforms) can be underlinked
+                    #      so it's better than an executable, as we'll only fail
+                    #      if it's unusable for another reason.
+
+                    # If we were asked specifically for a static library, don't perform
+                    # a link test, because it might be a special toolchain library.
+                    skip_link_check |= (libtype == LibType.STATIC)
+
+                    # Don't bother fishing for arguments if we've already
+                    # been told not to perform a test.
+                    if not skip_link_check:
+                        shared_link_args = self.get_std_shared_lib_link_args()
+                        extra_args = [trial] + lcargs + shared_link_args
+
+                        # Try to build the link test as PIC. We don't know if
+                        # the target library is itself PIC: if it's non-PIC,
+                        # PIC is fine, and if it is PIC, then PIC is needed,
+                        # so always use it.
+                        extra_args += self.get_pic_args()
+
+                        # Don't bother with a link check if we don't know how to build
+                        # a shared library.
+                        skip_link_check |= (not shared_link_args)
+
                     # When skip_link_check is True (e.g. for pkg-config
                     # libraries which are trusted to be linkable), skip the
                     # potentially expensive link check and just verify that the
                     # file exists.
-                    extra_args = [trial] + lcargs
                     if skip_link_check or self.links(code, extra_args=extra_args, disable_cache=True)[0]:
                         trial_result = trial
                         break

--- a/unittests/internaltests.py
+++ b/unittests/internaltests.py
@@ -721,6 +721,90 @@ class InternalTests(unittest.TestCase):
             env.machines.host.system = 'windows'
             self._test_all_naming(cc, patterns, 'windows-mingw')
 
+        self._test_find_library_undefined(cc)
+
+    def _test_find_library_undefined(self, cc):
+        '''
+        find_library checks if its argument both exists and can be
+        linked against, but it must tolerate underlinked static
+        libraries.
+
+        https://github.com/mesonbuild/meson/issues/15601
+        '''
+        def create_static_lib_with_undefined(name):
+            src = name.with_suffix('.c')
+            out = name.with_suffix('.o')
+            with src.open('w', encoding='utf-8') as f:
+                f.write('extern int get_cookie();')
+                f.write('int meson_foobar (void) { return get_cookie(); }')
+            # use of x86_64 is hardcoded in run_tests.py:get_fake_env()
+            if is_osx():
+                subprocess.check_call(['clang', '-c', str(src), '-o', str(out), '-arch', 'x86_64'])
+            else:
+                subprocess.check_call(['gcc', '-c', str(src), '-o', str(out)])
+            subprocess.check_call(['ar', 'csr', str(name), str(out)])
+
+        def create_static_lib_empty(name):
+            with name.open('w', encoding='utf-8') as f:
+                f.write("garbage")
+
+        # The test relies on some open-coded toolchain invocations for
+        # library creation in create_static_lib_with_undefined.
+        if is_windows() or is_cygwin():
+            return
+
+        with tempfile.TemporaryDirectory() as tmpdir:
+            p1 = Path(tmpdir)
+            create_static_lib_with_undefined(p1 / 'libfoo.a')
+
+            code = 'int meson_foobar (void); int main(void) { return meson_foobar(); }'
+
+            # Check that we always tolerate undefined references in a static
+            # library regardless of the library type.
+            for type in (LibType.STATIC, LibType.PREFER_STATIC, LibType.PREFER_SHARED):
+                found = cc._find_library_real('foo', [tmpdir],
+                                              code,
+                                              type, lib_prefix_warning=True, ignore_system_dirs=False)
+                self.assertEqual(os.path.basename(found[0]), 'libfoo.a')
+
+            # Check that we reject broken static libraries unless we were
+            # told to only find a static library.
+            create_static_lib_empty(p1 / 'libbar.a')
+
+            # We have a broken static library *and* we indicated we want a static
+            # library, so we don't perform a link test.
+            found = cc._find_library_real('bar', [tmpdir],
+                                          code,
+                                          LibType.STATIC, lib_prefix_warning=True, ignore_system_dirs=False)
+            self.assertEqual(os.path.basename(found[0]), 'libbar.a')
+
+            # We have a broken static library we're testing against but we only said
+            # we'd prefer static, not that it must be static: the heuristic
+            # says it likely isn't a special toolchain library, so we perform a
+            # link test.
+            found = cc._find_library_real('bar', [tmpdir],
+                                          code,
+                                          LibType.PREFER_STATIC, lib_prefix_warning=True, ignore_system_dirs=False)
+            self.assertIsNone(found, 'Unexpectedly found a library with PREFER_STATIC, link test expected to reject it')
+
+            # We have a broken static library we're testing against but we only said
+            # we'd prefer shared, not that it must be shared: the heuristic
+            # says it likely isn't a special toolchain library, so we perform a
+            # link test.
+            found = cc._find_library_real('bar', [tmpdir],
+                                          code,
+                                          LibType.PREFER_SHARED, lib_prefix_warning=True, ignore_system_dirs=False)
+            self.assertIsNone(found, 'Unexpectedly found a library with PREFER_SHARED, link test expected to reject it')
+
+            # We asked for a shared library and we only got a broken
+            # static one. We only tolerate them being broken if people
+            # explicitly ask for it w/ static: true, so we perform a link
+            # test.
+            found = cc._find_library_real('bar', [tmpdir],
+                                          code,
+                                          LibType.SHARED, lib_prefix_warning=True, ignore_system_dirs=False)
+            self.assertIsNone(found, "Unexpectedly found a library with SHARED")
+
     @skipIfNoPkgconfig
     def test_pkgconfig_parse_libs(self):
         '''

--- a/unittests/internaltests.py
+++ b/unittests/internaltests.py
@@ -729,6 +729,90 @@ class InternalTests(unittest.TestCase):
             env.machines.host.system = 'windows'
             self._test_all_naming(cc, patterns, 'windows-mingw')
 
+        self._test_find_library_undefined(cc)
+
+    def _test_find_library_undefined(self, cc):
+        '''
+        find_library checks if its argument both exists and can be
+        linked against, but it must tolerate underlinked static
+        libraries.
+
+        https://github.com/mesonbuild/meson/issues/15601
+        '''
+        def create_static_lib_with_undefined(name):
+            src = name.with_suffix('.c')
+            out = name.with_suffix('.o')
+            with src.open('w', encoding='utf-8') as f:
+                f.write('extern int get_cookie();')
+                f.write('int meson_foobar (void) { return get_cookie(); }')
+            # use of x86_64 is hardcoded in run_tests.py:get_fake_env()
+            if is_osx():
+                subprocess.check_call(['clang', '-c', str(src), '-o', str(out), '-arch', 'x86_64'])
+            else:
+                subprocess.check_call(['gcc', '-c', str(src), '-o', str(out)])
+            subprocess.check_call(['ar', 'csr', str(name), str(out)])
+
+        def create_static_lib_empty(name):
+            with name.open('w', encoding='utf-8') as f:
+                f.write("garbage")
+
+        # The test relies on some open-coded toolchain invocations for
+        # library creation in create_static_lib_with_undefined.
+        if is_windows() or is_cygwin():
+            return
+
+        with tempfile.TemporaryDirectory() as tmpdir:
+            p1 = Path(tmpdir)
+            create_static_lib_with_undefined(p1 / 'libfoo.a')
+
+            code = 'int meson_foobar (void); int main(void) { return meson_foobar(); }'
+
+            # Check that we always tolerate undefined references in a static
+            # library regardless of the library type.
+            for type in (LibType.STATIC, LibType.PREFER_STATIC, LibType.PREFER_SHARED):
+                found = cc._find_library_real('foo', [tmpdir],
+                                              code,
+                                              type, lib_prefix_warning=True, ignore_system_dirs=False)
+                self.assertEqual(os.path.basename(found[0]), 'libfoo.a')
+
+            # Check that we reject broken static libraries unless we were
+            # told to only find a static library.
+            create_static_lib_empty(p1 / 'libbar.a')
+
+            # We have a broken static library *and* we indicated we want a static
+            # library, so we don't perform a link test.
+            found = cc._find_library_real('bar', [tmpdir],
+                                          code,
+                                          LibType.STATIC, lib_prefix_warning=True, ignore_system_dirs=False)
+            self.assertEqual(os.path.basename(found[0]), 'libbar.a')
+
+            # We have a broken static library we're testing against but we only said
+            # we'd prefer static, not that it must be static: the heuristic
+            # says it likely isn't a special toolchain library, so we perform a
+            # link test.
+            found = cc._find_library_real('bar', [tmpdir],
+                                          code,
+                                          LibType.PREFER_STATIC, lib_prefix_warning=True, ignore_system_dirs=False)
+            self.assertIsNone(found, 'Unexpectedly found a library with PREFER_STATIC, link test expected to reject it')
+
+            # We have a broken static library we're testing against but we only said
+            # we'd prefer shared, not that it must be shared: the heuristic
+            # says it likely isn't a special toolchain library, so we perform a
+            # link test.
+            found = cc._find_library_real('bar', [tmpdir],
+                                          code,
+                                          LibType.PREFER_SHARED, lib_prefix_warning=True, ignore_system_dirs=False)
+            self.assertIsNone(found, 'Unexpectedly found a library with PREFER_SHARED, link test expected to reject it')
+
+            # We asked for a shared library and we only got a broken
+            # static one. We only tolerate them being broken if people
+            # explicitly ask for it w/ static: true, so we perform a link
+            # test.
+            found = cc._find_library_real('bar', [tmpdir],
+                                          code,
+                                          LibType.SHARED, lib_prefix_warning=True, ignore_system_dirs=False)
+            self.assertIsNone(found, 'Unexpectedly found a library with SHARED')
+
     @skipIfNoPkgconfig
     def test_pkgconfig_parse_libs(self):
         '''

